earth2studio/data/arco.py

```diff
@@ -207,30 +197,36 @@ async def create_data_array(
             },
         )
 
-        async def fetch_wrapper(
-            e: tuple[datetime, int, str, int],
-        ) -> tuple[int, int, np.ndarray]:
-            """Small wrapper that is awaitable for async generator"""
-            return e[1], e[3], self.fetch_array(e[0], e[2])
-
+        # Alternatives: https://death.andgravity.com/limit-concurrency
+        sem = asyncio.Semaphore(self.async_concurrency_limit)
         args = [
             (t, i, v, j) for j, v in enumerate(variable) for i, t in enumerate(time)
         ]
-        func_map = map(fetch_wrapper, args)
-
-        pbar = tqdm(
-            total=len(args), desc="Fetching ARCO data", disable=(not self._verbose)
+        func_map = map(
+            functools.partial(self.fetch_wrapper, xr_array=xr_array, sem=sem), args
         )
-        # Mypy will struggle here because the async generator uses a generic type
-        async for t, v, data in unordered_generator(  # type: ignore[misc,unused-ignore]
-            func_map, limit=self.async_process_limit
-        ):
-            xr_array[t, v] = data  # type: ignore[has-type,unused-ignore]
-            pbar.update(1)
 
+        # Before anything wait until the group gets opened
+        if self.zarr_major_version >= 3:
+            self.zarr_group = await self.zarr_group
+        # Launch all fetch requests
+        await tqdm.gather(
+            *func_map, desc="Fetching ARCO data", disable=(not self._verbose)
+        )
         return xr_array
 
-    def fetch_array(self, time: datetime, variable: str) -> np.ndarray:
+    async def fetch_wrapper(
+        self,
+        e: tuple[datetime, int, str, int],
+        xr_array: xr.DataArray,
+        sem: asyncio.Semaphore,
+    ) -> None:
+        """Small wrapper that is awaitable for async generator"""
+        async with sem:
```

Comment:
I'm not quire following why this is needed. Zarr v3 internally has concurrency limits too right? Also, you're not spawning any threads here, only using async right? So what issue is this semaphore here solving?

Reply:
This is fair, was trying to provide the user some control built in to rate limit how many requests are sent to a given api at once... but that can be done via zarr if needed. Removed.